### PR TITLE
Save why publishers and channels were not included in the payout report

### LINKED ISSUE #1618: Include all verified channels and publishers in payout report
Currently the payout report will only include verified channels and publishers that will receive payment.  As a result, there isn't an audit trail for publishers that were not paid out, and when publishers ask why they didn't get their payment, we don't have a document explaining why.  Ideally we should be tracking why each channel or publisher was not paid.  This will help us identify the source and scope of any bugs as well as provide better resources to our support team.

A verified channel / publisher will not be paid if 
* Their Uphold account `user status` is not `ok`, i.e. if it is `pending`, `restricted`, or `blocked`.
* They are not an Uphold member
* Brave is not authorized to read their Uphold account info
* They don't have a positive balance

Let's add these columns to the potential_payments table:
* `user_status` : string of (ok, pending, restricted, blocked)
* `was_member` : bool
* `reauthorization_needed` : bool

Instead of ignoring publishers that won't be paid out [here](https://github.com/brave-intl/publishers/blob/472654417873a56abc824a24c11423ac16339849/app/services/payout_report_publisher_includer.rb#L20) and [here](https://github.com/brave-intl/publishers/blob/472654417873a56abc824a24c11423ac16339849/app/services/payout_report_publisher_includer.rb#L35), let's create the potential payment for every verified channel and publisher. Then on the publisher page of the admin dashboard, an admin should be able to see a payment status section that displays why a publisher was or was not paid in the most recent final payout report.

### LINKED ISSUE #1631: Include uphold id in payout report

### PR DESCRIPTION
**Pull request by nvonpentz**
Resolves #1618 
Resolves #1631

Also, add uphold id to payout report.

Publisher page now displays report status information for the publisher and it's channels for the most recent final payout report:
image

Payout report page now displays the total amount of payments that are expected in the report as well as the actual number:
image

Submitter Checklist:

- [ ] Submitted a [ticket](https://github.com/brave-intl/publishers/issues) for my issue if one did not already exist.
- [ ] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [ ] Added/updated tests for this change (for new code or code which already has tests).
- [ ] Tagged reviewers.
- [ ] Integrated piwik/matomo (for code that adds new buttons).
- [ ] Addressed or ignored all brakeman warnings

Test Plan:

This can by tested by creating a publisher and a verified channel then generating payout reports with various states of suspension, uphold status, and balance amounts for the owner account and the channel.

Reviewer Checklist:

Tests
- [ ] Adequate test coverage exists to prevent regressions

Security:
- [ ] No raw SQL -- Always prefer ActiveRecord query helpers ([more info on StackOverflow](https://stackoverflow.com/questions/41410752/rails-5-sql-injection#41452695))
- [ ] XSS is mitigated -- Avoid `html_safe` and `raw`; escape untrusted content from users and 3rd party APIs ([Rails XSS guide](https://brakemanpro.com/2017/09/08/cross-site-scripting-in-rails) and [OWASP XSS Prevention Cheat Sheet](https://www.owasp.org/index.php/XSS_(Cross_Site_Scripting)_Prevention_Cheat_Sheet))

### CODE REVIEW

File: app/services/payout_report_publisher_includer.rb [hunk 1]
```diff
@@ -10,40 +10,52 @@ def perform
     publisher_has_unsettled_balance = false
 
     wallet = @publisher.wallet
-    # If the wallet has been blocked from uphold's terms of service. In this state users are unable to login or access uphold.
-    return if wallet.nil? || wallet.blocked?
```
Comment (nvonpentz) [inline app/services/payout_report_publisher_includer.rb]: The wallet should never be nil, and we want to record if it's blocked, so removing this line entirely

File: app/services/payout_report_publisher_includer.rb [hunk 2]
```diff
@@ -10,40 +10,52 @@ def perform
...
 
+    uphold_status = wallet.uphold_account_status
+    uphold_member = wallet.is_a_member?
+    uphold_connected = wallet.uphold_id.present? # TODO consider adding sub category for when wallet.action == 're-authorize'
```
Comment (nvonpentz) [inline app/services/payout_report_publisher_includer.rb]: I realized this is not needed and can also be expressed as `uphold_account_status == nil`, I will track reauthorization need instead.

File: app/helpers/admin_helper.rb [hunk 3]
```diff
@@ -15,4 +15,14 @@ def nav_link(text, path)
       link_to text, path, options
     end
   end
+
+  def payout_report_status_header(account_type)
+    report_date = PayoutReport.most_recent_final_report.created_at.strftime("%b %d")
```
Comment (yachtcaptain23) [inline app/helpers/admin_helper.rb]: Nit picking: Can we do `%Y-%m-%d`?

File: db/schema.rb [hunk 4]
```diff
@@ -235,6 +236,10 @@
     t.string "amount", null: false
     t.string "fees", null: false
     t.string "url"
+    t.string "uphold_status_was"
+    t.boolean "reauthorization_was_needed", null: false
```
Comment (yachtcaptain23) [inline db/schema.rb]: Would you update the schema to make sure `null: false` isn't included?

File: db/migrate/20190219194426_add_report_status_info_to_potential_payments.rb [hunk 5]
```diff
@@ -0,0 +1,8 @@
+class AddReportStatusInfoToPotentialPayments < ActiveRecord::Migration[5.2]
+  def change
+    add_column :potential_payments, :uphold_status_was, :string
```
Comment (yachtcaptain23) [inline db/migrate/20190219194426_add_report_status_info_to_potential_payments.rb]: IMO you should drop `_was` in all these columns. It's fairly self explanatory it's a snapshot. 

@corymcdonald @dlipeles do you agree?

File: app/services/payout_report_publisher_includer.rb [hunk 6]
```diff
@@ -10,40 +10,52 @@ def perform
...
+    uphold_status = wallet.uphold_account_status
+    uphold_member = wallet.is_a_member?
+    reauthorization_needed = wallet.action == "re-authorize"
+    suspended = @publisher.suspended?
+
+    # Create potential payment for referrals
     probi = wallet.channel_balances[@publisher.owner_identifier].probi_before_fees # probi = balance
-    if probi.positive?
-      publisher_has_unsettled_balance = true
+    publisher_has_unsettled_balance = probi.positive? ? true : false
```
Comment (yachtcaptain23) [inline app/services/payout_report_publisher_includer.rb]: `publisher_has_unsettled_balance = probi.positive?`

File: db/schema.rb (hunk 4 above)
Comment (nvonpentz) [reply db/schema.rb]: good catch!

File: app/helpers/admin_helper.rb (hunk 3 above)
Comment (nvonpentz) [reply app/helpers/admin_helper.rb]: I agree that your format is better for most things, but in this case I wonder if it makes it less readable.  The date here should really only be from the current month if everything is working, so I think including the year might be overkill.

image
vs
image

@corymcdonald @dlipeles do you have an opinion?

File: app/services/payout_report_publisher_includer.rb (hunk 6 above)
Comment (nvonpentz) [reply app/services/payout_report_publisher_includer.rb]: Resolved in 023ddd4

File: app/helpers/admin_helper.rb (hunk 3 above)
Comment (dlipeles) [reply app/helpers/admin_helper.rb]: BD has been asking us to make things more user friendly so I'll vote for month/day

File: db/migrate/20190219194426_add_report_status_info_to_potential_payments.rb (hunk 5 above)
Comment (corymcdonald) [reply db/migrate/20190219194426_add_report_status_info_to_potential_payments.rb]: aye 👍

Comment (nvonpentz) [reply db/migrate/20190219194426_add_report_status_info_to_potential_payments.rb]: thanks guys, will update

**Comment by nvonpentz** [comment]
This now also resolves #1631

File: app/helpers/admin_helper.rb (hunk 3 above)
Comment (nvonpentz) [reply app/helpers/admin_helper.rb]: I will keep it this way unless you want to veto @yachtcaptain23

File: db/migrate/20190219194426_add_report_status_info_to_potential_payments.rb (hunk 5 above)
Comment (nvonpentz) [reply db/migrate/20190219194426_add_report_status_info_to_potential_payments.rb]: fb4c7d7

**Pull request merged**